src/Backend/Haskell.idr

```diff
@@ -117,22 +347,386 @@ mutual
+decode t@(TApp f xs) = decoderTerm t -- assumes the def of this is generated elsewhere
+
+||| Generate an encoder function definition for the given TNamed.
+||| Assumes definitions it depends on are generated elsewhere.
+encodeDef : TNamed n -> Haskell
+encodeDef {n} t@(TName tname td) =
+  let env = freshEnvWithTerms "encode"
+      envTypes = map fst env
+      funName = if tname == ""
+                then encodeName (makeType envTypes td)
+                else "encode" ++ uppercase tname
+      varEncs = toList $ getUsedVars (map snd env) td
+      currTerm = HsApp (HsFun funName) varEncs
+      currType = if tname == ""
+                 then makeType envTypes td
+                 else HsParam tname []
+      funType = foldr HsArrow (hsSerialiser (makeType' envTypes t)) (map hsSerialiser (getUsedVars envTypes td))
+      clauses = genClauses n currType currTerm env varEncs td in
+      FunDef funName funType clauses
+  where genConstructor : (n : Nat) -> String -> TDef n -> TermGen n (HsTerm, List HsTerm)
+        genConstructor n name (TProd {k = k} xs) = do
+          xs' <- freshVars (2 + k) "x"
+          rhs <- mapWithIndexA (\ i, t' => encode (index i xs) t') xs'
+          pure $ (HsInn name (toList xs'), toList rhs)
+        genConstructor n name td = do
+          [x'] <- freshVars 1 "x"
+          rhs <- encode td x'
+          pure $ (HsInn name [x'], [rhs])
+
+        genClause : (n : Nat) -> List HsTerm -> Fin k -> (String, TDef n) -> TermGen n (List HsTerm, HsTerm)
+        genClause n varEncs i (name, T1) = pure (varEncs ++ [HsInn name []], word (fromInteger (finToInteger i)))
+        genClause n varEncs i (name, args) = do
+          (con, rhs) <- genConstructor n name args
+          pure (varEncs ++ [con], simplify $ HsConcat (word (fromInteger (finToInteger i))::rhs))
+
+        -- Idris is not clever enough to figure out the following type if written as a case expression
+        genClauses : (n : Nat) -> HsType -> HsTerm -> Vect n (HsType, HsTerm) -> List HsTerm -> TDef n -> List (List HsTerm, HsTerm)
+        genClauses n currType currTerm env varEncs (TMu [(name, td)]) = runTermGen ((currType, currTerm)::env) $ do
+          (con, rhs) <- genConstructor (S n) name td
+          pure [(varEncs ++ [con], simplify $ HsConcat rhs)]
+        genClauses n currType currTerm env varEncs (TMu tds) = toList $ runTermGen ((currType, currTerm)::env) (mapWithIndexA (genClause (S n) varEncs) tds)
+        genClauses n currType currTerm env varEncs td        = let v = HsTermVar "x" in [( varEncs ++ [v] , simplify $ runTermGen env (encode td v))]
+
+||| Generate an decoder function definition for the given TNamed.
+||| Assumes definitions it depends on are generated elsewhere.
+decodeDef : TNamed n -> Haskell
+decodeDef {n = n} t@(TName tname td) =
+  let env = freshEnvWithTerms "decode"
+      envTypes = map fst env
+      funName = if tname == ""
+                then decodeName (makeType envTypes td)
```

Comment:
Same comment as above, why do we not use `decode ++ _type name_` ?

Reply:
Again, we could.